### PR TITLE
feat(platform-webworker): renderer v2 integration

### PR DESCRIPTION
**Pull request by vicb**

### CODE REVIEW

File: modules/@angular/platform-webworker/src/web_workers/worker/renderer.ts [hunk 1]
```diff
@@ -6,25 +6,23 @@
...
 
-import {deserializeGenericEvent} from './event_deserializer';
-
 @Injectable()
 export class WebWorkerRootRenderer implements RootRenderer {
-  private _messageBroker: any /** TODO #9100 */;
-  public globalEvents: NamedEventEmitter = new NamedEventEmitter();
-  private _componentRenderers: Map<string, WebWorkerRenderer> =
-      new Map<string, WebWorkerRenderer>();
+  globalEvents = new NamedEventEmitter();
```
Comment (vicb) [inline modules/@angular/platform-webworker/src/web_workers/worker/renderer.ts]: delete

File: tools/public_api_guard/platform-webworker/index.d.ts [hunk 2]
```diff
@@ -61,6 +61,11 @@ export declare class ReceivedMessage {
 }
 
 /** @experimental */
+export declare const enum SerializerTypes {
+    RENDERER_TYPE_V2 = 0,
```
Comment (vicb) [inline tools/public_api_guard/platform-webworker/index.d.ts]: move `PRIMITIVE` here

File: modules/@angular/platform-webworker/src/web_workers/shared/client_message_broker.ts [hunk 3]
```diff
@@ -100,74 +100,59 @@ export class ClientMessageBroker_ extends ClientMessageBroker {
       promise = new Promise((resolve, reject) => { completer = {resolve, reject}; });
       id = this._generateMessageId(args.method);
       this._pending.set(id, completer);
+
       promise.catch((err) => {
-        print(err);
+        if (console && console.log) {
+          // tslint:disable-next-line:no-console
+          console.log(err);
```
Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/shared/client_message_broker.ts]: should this be `console.error`?

File: modules/@angular/platform-webworker/src/web_workers/shared/client_message_broker.ts [hunk 4]
```diff
@@ -100,74 +100,59 @@ export class ClientMessageBroker_ extends ClientMessageBroker {
...
-      promise = promise.then((value: any) => {
-        if (this._serializer == null) {
-          return value;
-        } else {
-          return this._serializer.deserialize(value, returnType);
-        }
-      });
+      promise = promise.then(
+          (value: any) =>
+              this._serializer ? value : this._serializer.deserialize(value, returnType));
```
Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/shared/client_message_broker.ts]: should be `this._serializer ? this._serializer.deserialize(value, returnType) : value` (the cases are in the wrong order)

File: modules/@angular/platform-webworker/src/web_workers/shared/client_message_broker.ts [hunk 5]
```diff
@@ -100,74 +100,59 @@ export class ClientMessageBroker_ extends ClientMessageBroker {
...
     } else {
       promise = null;
     }
 
-    // TODO(jteplitz602): Create a class for these messages so we don't keep using StringMap #3685
     const message = {'method': args.method, 'args': fnArgs};
     if (id != null) {
-      (message as any /** TODO #9100 */)['id'] = id;
+      (message as any)['id'] = id;
```
Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/shared/client_message_broker.ts]: maybe introduce a `RequestMessageData`? Then you can type the object above and there is not need for `message as any`... And maybe rename `MessageData` below into `ResponseMessageData`.

File: modules/@angular/platform-webworker/src/web_workers/shared/render_store.ts [hunk 6]
```diff
@@ -33,21 +28,8 @@ export class RenderStore {
...
-    if (id == null) {
-      return null;
-    }
-
-    if (!this._lookupById.has(id)) {
-      return null;
-    }
-
-    return this._lookupById.get(id);
+    return id == null || !this._lookupById.has(id) ? null : this._lookupById.get(id);
```
Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/shared/render_store.ts]: This is a bit too much nesting for my taste...
Please add one `if` for the check `id == null`.

File: modules/@angular/platform-webworker/src/web_workers/shared/serializer.ts [hunk 7]
```diff
@@ -6,79 +6,87 @@
...
-// PRIMITIVE is any type that does not need to be serialized (string, number, boolean)
-// We set it to String so that it is considered a Type.
+
 /**
+ * Any type that does not need to be serialized (string, number, boolean)
+ *
  * @experimental WebWorker support in Angular is currently experimental.
+ * @deprecated in v4. Use SerializerTypes.PRIMITIVE instead
  */
 export const PRIMITIVE: Type<any> = String;
```
Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/shared/serializer.ts]: why not set the value of this to `SerializerTypes.PRIMITIVE`?

File: modules/@angular/platform-webworker/src/web_workers/shared/serializer.ts [hunk 8]
```diff
@@ -6,79 +6,87 @@
...
+      public port: string, public pathname: string, public search: string, public hash: string,
+      public origin: string) {}
+}
+
+/**
+ * @experimental WebWorker support in Angular is currently experimental.
+ */
+export const enum SerializerTypes {
+  // RendererTypeV2
+  RENDERER_TYPE_V2,
```
Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/shared/serializer.ts]: Why a special enum here, and not use the `RendererV2` class as type value? I.e. similar to `LocationType` and the `Renderer` (v1)... `RendererV2` is now an abstract class, so this should work...

Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/shared/serializer.ts]: Ah, I meant `RendererTypeV2`. And we should then make it an abstract class...

Comment (vicb) [inline modules/@angular/platform-webworker/src/web_workers/shared/serializer.ts]: `export interface RendererTypeV2`

Comment (vicb) [inline modules/@angular/platform-webworker/src/web_workers/shared/serializer.ts]: I would prefer an if is there is no other justification to make it an abstract class. WDYT ?

File: modules/@angular/platform-webworker/test/web_workers/shared/web_worker_test_util.ts [hunk 9]
```diff
@@ -42,7 +42,7 @@ export function createPairedMessageBuses(): PairedMessageBuses {
 export function expectBrokerCall(
     broker: SpyMessageBroker, methodName: string, vals?: Array<any>,
     handler?: (..._: any[]) => Promise<any>| void): void {
-  broker.spy('runOnService').and.callFake((args: UiArguments, returnType: Type<any>) => {
+  broker.spy('callUI').and.callFake((args: UiArguments, returnType: Type<any>) => {
```
Comment (tbosch) [inline modules/@angular/platform-webworker/test/web_workers/shared/web_worker_test_util.ts]: I think this should stay `runOnService`?

File: modules/@angular/platform-webworker/src/web_workers/worker/renderer.ts [hunk 10]
```diff
@@ -6,25 +6,54 @@
...
+    let listeners = this._listeners.get(eventName);
+    if (!listeners) {
+      listeners = [];
+      this._listeners.set(eventName, listeners);
+    }
+    return listeners;
+  }
+}
+
+const globalEvents = new NamedEventEmitter();
```
Comment (tbosch) [inline modules/@angular/platform-webworker/src/web_workers/worker/renderer.ts]: Why are these outside of the `WebWorkerRootRenderer` class? This way, they might leak when a user bootstraps multiple times...

**Review by tbosch — APPROVED** [review]

**Comment by tbosch** [comment]
Only minor nits.

File: modules/@angular/platform-webworker/test/web_workers/shared/web_worker_test_util.ts (hunk 9 above)
Comment (vicb) [inline modules/@angular/platform-webworker/test/web_workers/shared/web_worker_test_util.ts]: yep, the IDE was probably too smart here

**Pull request closed without merging**